### PR TITLE
Add GetUnitsByProperty request

### PR DESCRIPTION
**Pull request by tcollier**
@kristjan @heidi @jrichard

Added `GetUnitsByProperty` request and supporting objects, namely the public `Unit` model and `ListCriterion` wrapper for the `list_criteria` argument.

I've also added a DSL for declaring attributes by data type for models. This added the `AttributeParser`, which prompted renaming the `Parser` module to `DocumentParser` for clarity.

<a href="https://github.com/apartmentlist/real_page/tree/get-units-by-property#realpage">Updated README</a>.

### CODE REVIEW

File: spec/support/shared_examples/model.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  it "casts #{attr} to a boolean" do
+    subject = described_class.new(attr => '1')
+    expect(subject.send("#{attr}?")).to eq(true)
+  end
+end
+
+shared_examples_for 'a date attribute' do |attr|
+  it "casts #{attr} to a date" do
+    subject = described_class.new(attr => '5/5/2015')
+    expect(subject.send(attr)).to eq(Date.parse('2015-05-05'))
```
Comment (jrichard) [inline spec/support/shared_examples/model.rb]: parsed value differs from value set on attr above, should be the same.

Comment (jrichard) [inline spec/support/shared_examples/model.rb]: ... on second thought, these shared examples are really showing that type coercion from XML string to ruby types occurs.  The individual AttributeParser tests show parsing correctness.  Should these specs just assert on the output type instead?

Comment (tcollier) [inline spec/support/shared_examples/model.rb]: :+1:

**Comment by jrichard** [comment]
lgtm

**Pull request merged**